Themes: Add theme download in sheet view.

mendezcode opened the pull request:
Implements theme downloads into sheet view as specified in #3660.

Tracking of downloads coming down the road.

Premium themes are not downloaded, therefore not showing the card on those.

There are cases, in which themes do not have `download_uri` set. Such as the case with Twenty Sixteen, which recommends downloading from WordPress.org. In this particular case, the card is shown like this:

![screen shot 2016-03-31 at 4 26 46 pm](https://cloud.githubusercontent.com/assets/5881557/14189892/bf2c271a-f75d-11e5-8ff1-421b742dcc32.png)

The download button (and link in the description) will point to `http://wordpress.org/themes/twentysixteen`.

The text above has been adapted from the original Twenty Sixteen page from the theme showcase: https://theme.wordpress.com/themes/twentysixteen/

Feedback and suggestions welcome!

seear commented on client/my-sites/themes/sheet.jsx:
```diff
@@ -134,6 +144,32 @@ const ThemeSheet = React.createClass( {
+			return '';
+		}
+		downloadText = this.props.download
+		? i18n.translate( 'This theme is available for download to be used on your {{u}}WordPress self-hosted{{/u}} installation.', {
```
I think the underlined part in the design is intended to be a link to somewhere @folletto ?

folletto replied:
Yep, should be something specific, but initially the wordpress.org site is the starting point. :)

seear commented:
Instead of a button click, we may be able to have a direct link to the zip, styled as a button, which would avoid the extra browser tab on click.

seear commented:
We could also remove or disable the button in the case where we link to the dot org download. Or have the button be the dot org link. cc @folletto.

seear commented on client/my-sites/themes/style.scss:
```diff
@@ -253,3 +273,30 @@
 	}
 }
+
+@include breakpoint( ">1040px" ) {
```
Any reason that this is not a `>660px` breakpoint?

mendezcode replied:
The theme screenshot column takes some space on the right, causing the left column to be a bit narrow.

I used `>1040px` to avoid this behavior:

![screen shot 2016-04-24 at 11 51 23 am](https://cloud.githubusercontent.com/assets/5881557/14768722/a73d0c7e-0a17-11e6-81cf-2ad09c858034.png)

I'm not sure if it's OK to have this though? Thoughts?

folletto replied:
It's ok. One day we'll be able to do that contextual on the actual width of the component instead of the global viewport width, but for today... I think it's effective. Not ideal, but hey, works. Maybe let's add a one-line comment in the code to explain why, given the question might come up:

> Tweaked breakpoint because this component is used in a two-columns layout and gets narrower earlier.

seear commented:
@mendezcode: you may want to experiment with having the download box as a separate react component, to keep the sheet component tidy. There is nothing wrong with the way it currently is, though, so that is entirely up to you.

folletto commented:
> having the download box as a separate react component

I'd add that this would be ideal just in case we'd want to enable later this feature in other parts of the UI :)

ockham commented:
> Instead of a button click, we may be able to have a direct link to the zip, styled as a button, which would avoid the extra browser tab on click.

:+1: The `Button` component supports that out-of-the-box, BTW -- see the [`href` prop](https://github.com/Automattic/wp-calypso/tree/master/client/components/button#props).

folletto commented:
> There are cases, in which themes do not have download_uri set. Such as the case with Twenty Sixteen, which recommends downloading from WordPress.org. In this particular case, the card is shown like this

Ok two things here:

1. Can we keep the text the same? I'm not sure there's any value in specifying the source in that situation.
2. Can we link directly to the download from .org? 

Also: how do we know that's the right URL, and that URL should be used? What's the condition?

mendezcode commented:
> Instead of a button click, we may be able to have a direct link to the zip, styled as a button, which would avoid the extra browser tab on click.

@seear Agreed! using `href` makes more sense. I have refactored accordingly.

> you may want to experiment with having the download box as a separate react component, to keep the sheet component tidy.

@seear That's right! This has been refactored and it is considerably more elegant indeed. Thanks for the suggestion!

Also, thanks @ockham for suggesting the `<Button />` component :)

---

@folletto re: your questions:

> 1. Can we keep the text the same? I'm not sure there's any value in specifying the source in that situation.

Yes, doesn't actually make sense to have two different texts. Rookie mistake, fixed!

> 2. Can we link directly to the download from .org?

I was confused initially since the .org version of the downloads included the version number, which we do not have in wpcom. But, after some testing I did realize we can donwload the theme without the version number.

I changed it to include the direct link from .org, sorry about the confusion!

folletto commented on client/my-sites/themes/theme-download-card/index.jsx:
```diff
@@ -21,7 +21,7 @@ var ThemeDownloadCard = React.createClass( {
 
 	render() {
-		var downloadURI = this.props.href || ( 'http://wordpress.org/themes/' + this.props.theme );
+		var downloadURI = this.props.href || ( 'https://downloads.wordpress.org/theme/' + this.props.theme + '.zip' );
```
Does this mean that if it's empty we assume it's on .org? Is this a fair assumption to make?

folletto commented on client/my-sites/themes/sheet.jsx:
```diff
@@ -51,6 +53,14 @@ const ThemeSheet = React.createClass( {
 	},
 
+	onDownloadButtonClick() {
+		if ( this.props.download ) {
```
Does this mean that if it's empty we assume it's on .org? Is it a fair assumption to make?

mendezcode replied:
This is a tricky question, usually we create we create ZIPs for the themes we convert. In the case of TwentySixteen though, there was no download generated, so I assume when this happens the theme is being directed to be downloaded from WordPress.org.

This can be seen in the Twenty Sixteen showcase page: 

https://theme.wordpress.com/themes/twentysixteen/

folletto replied:
Ok so we can start there, and we'll review later once we have a better API-driven way to determine this?

seear commented:
Looks like `onDownloadButtonClick() ` in sheet.jsx can be removed completely now.

mendezcode commented:
> Looks like `onDownloadButtonClick()` in sheet.jsx can be removed completely now.

Missed that one... Thanks for spotting!

Will refactor accordingly in the next commit.

folletto commented on client/my-sites/themes/theme-download-card/index.jsx:
```diff
@@ -0,0 +1,40 @@
+	},
+
+	render() {
+		var downloadURI = this.props.href || ( 'https://downloads.wordpress.org/theme/' + this.props.theme + '.zip' );
```
Let's add here an inline code comment explaining why we do fallback by default to ".org". ;)

Like:

```// When we don't generate zips, it's because we have released the theme on .org.```

Maybe?

mendezcode replied:
Makes sense yes.

seear commented on client/my-sites/themes/theme-download-card/index.jsx:
```diff
@@ -0,0 +1,40 @@
+
+	render() {
+		var downloadURI = this.props.href || ( 'https://downloads.wordpress.org/theme/' + this.props.theme + '.zip' );
+		var downloadText = i18n.translate( 'This theme is available for download to be used on your {{a}}WordPress self-hosted{{/a}} installation.', {
```
The [preferred style](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#let-and-const) is to use `const` (or `let` when necessary) over `var`. For example: https://github.com/Automattic/wp-calypso/blob/d5360c590fa32f32bc226761d043c64dd32fc13d/client/my-sites/themes/sheet.jsx#L73

seear commented on client/my-sites/themes/theme-download-card/index.jsx:
```diff
@@ -0,0 +1,40 @@
+		var downloadURI = this.props.href || ( 'https://downloads.wordpress.org/theme/' + this.props.theme + '.zip' );
+		var downloadText = i18n.translate( 'This theme is available for download to be used on your {{a}}WordPress self-hosted{{/a}} installation.', {
+			components: {
+				a: <a href={ downloadURI } />
```
I think maybe the should be something like https://wordpress.org/ rather than the download link (see https://github.com/Automattic/wp-calypso/pull/4473#discussion_r58350369)

seear commented:
> I think maybe the should be something like https://wordpress.org/ rather than the download link 

Sorry @mendezcode, you had the button link exactly right—I was referring to the _WordPress.org self-hosted_ link. I've tweaked it.

I'll rebase this PR because we've moved some stuff around in the meantime, then it will be good to merge.

seear commented:
@mendezcode: I had to squash all the commits to make the rebase easier. This is ready to merge whenever you want. Excellent work again 👍
